File: packages/core/src/retry.ts
/**
 * Delivery retry classification shared by browser queue and server senders.
 *
 * At-most-once semantics: senders may retry only RETRYABLE outcomes while
 * reusing the original event_id; collectors no-op duplicates.
 */
export type DeliveryOutcome =
  | { kind: 'delivered' }
  | { kind: 'retryable'; reason: string }
  | { kind: 'permanent'; reason: string };

/** HTTP status -> retry classification for collector deliveries. */
export function classifyDeliveryStatus(status: number): DeliveryOutcome {
  if (status >= 200 && status < 300) return { kind: 'delivered' };
  if (status === 408 || status === 425 || status === 429) {
    return { kind: 'retryable', reason: `http_${status}` };
  }
  if (status >= 500) return { kind: 'retryable', reason: `http_${status}` };
  return { kind: 'permanent', reason: `http_${status}` };
}

/** Network-level failure (no status): always retryable at the sender's policy. */
export function classifyNetworkError(): DeliveryOutcome {
  return { kind: 'retryable', reason: 'network_error' };
}

File: python/clicktrail/clicktrail/retry.py
"""Delivery retry classification shared by server senders (JS core parity).

At-most-once semantics: senders may retry only RETRYABLE outcomes while
reusing the original event_id; collectors no-op duplicates.
"""

from typing import NamedTuple, Optional


class DeliveryOutcome(NamedTuple):
    kind: str  # 'delivered' | 'retryable' | 'permanent'
    reason: Optional[str] = None


def classify_delivery_status(status: int) -> DeliveryOutcome:
    """Map an HTTP status to a delivery classification."""
    if 200 <= status < 300:
        return DeliveryOutcome("delivered")
    if status in (408, 425, 429):
        return DeliveryOutcome("retryable", "http_%d" % status)
    if status >= 500:
        return DeliveryOutcome("retryable", "http_%d" % status)
    return DeliveryOutcome("permanent", "http_%d" % status)


def classify_network_error() -> DeliveryOutcome:
    """Network-level failure (no status): always retryable at sender policy."""
    return DeliveryOutcome("retryable", "network_error")
